net: tc: Priority to Traffic Class mapping improvements

tgorochowik opened the pull request:
This commits adds new priority to traffic class mappings and allows
users to choose which mapping to use through menuconfig.

The new mappings are recommended in 802.1 (chapter 33.5) for
time-sensitive applications supporting the credit-based sharper
algorithm.

Please review, thanks!

tgorochowik committed:
net: tc: Add other priority to traffic class mappings

This commits adds new priority to traffic class mappings and allows
users to choose which mapping to use through menuconfig.

The new mappings are recommended in 802.1 (chapter 34.5) for
time-sensitive applications supporting the credit-based sharper
algorithm.

Signed-off-by: Tomasz Gorochowik <tgorochowik@antmicro.com>

tgorochowik committed:
net: tc: Fix priority to traffic class mappings

Priority 0 is the default, but 1 is the lowest. See 802.1Q tables I-1
and I-2.

Signed-off-by: Tomasz Gorochowik <tgorochowik@antmicro.com>

jukkar reviewed (changes requested):
Minor tweak needed, otherwise ok.

jukkar commented:
We also need to test these new mappings. Different combinations of the new mappings need to be added to tests/net/traffic_class/testcase.yaml file. Not sure if the actual unit test would work as is or does it need modifications.

jukkar commented on subsys/net/ip/net_tc_mapping.h:
```diff
@@ -0,0 +1,140 @@
+ * according to 802.1Q - table I-2.
+ *
+ *  Priority         Acronym   Traffic types
+ *  1 (lowest)       BK        Background
```
Please fix also the NET_PRIORITY_xx settings in net_ip.h

tgorochowik replied:
Added a commit that does just that.

tgorochowik committed:
tests: net: Add new TC mappings testcases

The new mappings require at least two traffic classes so there are fewer
tests than there are for the strict mapping.

Signed-off-by: Tomasz Gorochowik <tgorochowik@antmicro.com>

tgorochowik committed:
net: ip: Fix net priorities definitions

Best Effort is the default priority with the assigned value of 0, but
Background is the lowest priority with the assigned value of 1.

Ref: IEEE 802.1Q, Chapter I.4, Table I-2.

Signed-off-by: Tomasz Gorochowik <tgorochowik@antmicro.com>

tgorochowik commented:
@jukkar Thank you for your review!

The maps maps have been changed a little bit as in the previous implementation it did not take into account that there could be a different number of TX and RX TC classes. Is the current implementation ok? With it there is no need to duplicate the maps for RX and TX.

The issues you raised in your review are addressed - there is just one thing to verify - I am not sure if the PCP (vlan priority) to Packet priority conversion is still needed.

It looks like both the PCP values and Packet Priorities values are swapped (BE and BK priorities). So earlier we did fix VLAN priorities conversion, but it seems that not in a proper way. This standard is just so confusing when it comes to priorities.

jukkar commented on subsys/net/ip/net_tc.c:
```diff
@@ -52,8 +52,8 @@ int net_tx_priority2tc(enum net_priority prio)
 	 * chapter 34.5 table 34-1
```
The commit message should probably say "...but 1 is the lowest."

jukkar reviewed (approved):
LGTM

jukkar reviewed (changes requested):
Forgot that the commit message needs still fixing.

pfalcon commented on subsys/net/ip/Kconfig:
```diff
@@ -70,6 +70,39 @@ config NET_TC_RX_COUNT
+	help
+	  This is the recommended default priority to traffic class mapping.
+	  Use it for implementations that do not support the credit-based
+	  sharper transmission selection algorithm.
```
@tgorochowik: Post-review: a typo sneaked in, s/sharper/shaper/. Hope you can see this and submit a patch ;-).

pfalcon commented on subsys/net/ip/net_tc_mapping.h:
```diff
@@ -0,0 +1,138 @@
+#if defined(CONFIG_NET_TC_MAPPING_STRICT)
+
+/* This is the recommended priority to traffic class mapping for
+ * implementations that do not support the credit-based sharper transmission
```
Also "sharper". Probably worth doing search-and-replace...

tgorochowik replied:
@pfalcon Right! Sorry about that! This is actually a copy of the same text (Header and Kconfig), thank you for letting me know :+1: I'll submit a patch for that.
